Sources/CoreFoundation/CFString.c

```diff
@@ -1294,6 +1294,12 @@ CF_PRIVATE CFStringRef __CFStringCreateImmutableFunnel3(
                         Boolean possiblyExternalFormat, Boolean tryToReduceUnicode, Boolean hasLengthByte, Boolean hasNullByte, Boolean noCopy,
                         CFAllocatorRef contentsDeallocator, UInt32 converterFlags) {
     
+    if (hasNullByte && !__CFStringEncodingIsSupersetOfASCII(encoding)) {
+        // Non-8bit encodings cannot be safely read as c-strings because they may contain many null bytes
+        // This was documented as invalid previously, but now we validate that eagerly here to prevent creating truncated strings or strings that incorrectly assume 8bit representation
+        HALT_MSG("CFStringCreateWithCString can only be called with an 8-bit encoding");
```

Comment:
Do you think it might be possible that people are calling this with some non-8bit encoding, but only ASCII data, and getting away with it today?

Reply:
For many encodings I suspect they couldn't get away with it (for example in UTF-16 ASCII text would have a lot of intermediate null bytes so depending on the endianness either string creation would fail (because there would be an odd number of bytes) or the string would be empty (because the first byte is a null byte)). So it's possible that someone could be relying on the failing (`NULL` return) or empty string behavior, but I think that's not super likely. It's possible that there's another encoding that we have that might work - I can double check - but I can't think of one off the top of my head that would behave correctly